### PR TITLE
[Ready] logsumexp for raggedtensor

### PR DESCRIPTION
**Pull request by glynpu**

### CODE REVIEW

File: k2/python/tests/ragged_tensor_test.py [hunk 1]
```diff
@@ -227,6 +228,55 @@ def test_sum_no_grad(self):
...
+
+                b = a.logsumexp()
+                b_0 = a_0.logsumexp(dim=0)
+                b_1 = a_1.logsumexp(dim=0)
+                b_2 = a_2.logsumexp(dim=0)
+
+                expected_b = torch.stack([b_0, b_1, b_2])
+                assert torch.allclose(b, expected_b)
+
+                c = b.sum()
```
Comment (csukuangfj) [inline k2/python/tests/ragged_tensor_test.py:254]: Could you also test the scale case, e.g.
```python3
c = (b * torch.arange(b.numel()).to(b)).sum()
```

File: k2/python/tests/ragged_tensor_test.py [hunk 2]
```diff
@@ -227,6 +228,55 @@ def test_sum_no_grad(self):
...
+                expected_c = expected_b.sum()
+                c.backward()
+                expected_c.backward()
+                assert torch.allclose(a.grad,
+                                      torch.cat([a_0.grad, a_1.grad, a_2.grad]))
+
+    def test_logsumexp_no_grad(self):
```
Comment (csukuangfj) [inline k2/python/tests/ragged_tensor_test.py:261]: This function can be merged with the above function to reduce the amount of test code.

File: k2/python/csrc/torch/v2/doc/any.h [hunk 3]
```diff
@@ -930,6 +930,43 @@ tensor(40., grad_fn=<SumBackward0>)
...
+Note:
+  This operation only supports float type input,
+  i.e., with dtype being torch.float32 or torch.float64.
+
+>>> import torch
+>>> import k2
+>>> import k2.ragged as k2r
+>>> a = k2r.RaggedTensor([[-0.25, -0.25, -0.25, -0.25], [], [-0.5, -0.5]], dtype=torch.float32)
+>>> a.requires_grad_(True)
+RaggedTensor([[-0.25, -0.25, -0.25, -0.25],
```
Comment (csukuangfj) [inline k2/python/csrc/torch/v2/doc/any.h:949]: Could you also add an example for ragged tensors with 3-axis?

File: k2/python/csrc/torch/v2/autograd/logsumexp.h [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+      if (std::is_same<T, float>::value) {
+        // use `expf` for float
+        K2_EVAL(
+            c, n, set_grad, (int32_t idx01)->void {
+              int32_t idx0 = row_ids_data[idx01];
+              ans_data[idx01] = grad_output_data[idx0 * stride]
+                * expf(input_logp_data[idx01] - logsum_values_data[idx0]);
+            });
+      } else {
+        // use `exp` foradouble
```
Comment (csukuangfj) [inline k2/python/csrc/torch/v2/autograd/logsumexp.h:114]:
```suggestion
        // use `exp` for double
```

File: k2/python/tests/ragged_tensor_test.py [hunk 5]
```diff
@@ -227,6 +228,80 @@ def test_sum_no_grad(self):
...
 
+    def test_logsumexp_with_grad(self):
+        for device in self.devices:
+            for dtype in self.real_dtypes:
+                a = k2r.RaggedTensor(
+                    [[-1.1, 2.2, -3.3, 4.4], [], [5.5, -6.6]],
+                    dtype=dtype,
+                    device=device,
+                )
+                a_0 = torch.tensor(
```
Comment (csukuangfj) [inline k2/python/tests/ragged_tensor_test.py:239]: I suggest that you use
```python3
a_0 = a[0].clone()
```

(also for a_1 a_2)

**Review by csukuangfj — CHANGES_REQUESTED** [review]

File: k2/python/csrc/torch/v2/doc/any.h [hunk 6]
```diff
@@ -930,6 +930,64 @@ tensor(40., grad_fn=<SumBackward0>)
   containing the computed sum.
 )doc";
 
+static constexpr const char *kRaggedAnyLogSumExpDoc = R"doc(
+Compute the logsumexp of sublists over the last axis of this tensor.
+
+Note:
+  It is similar to torch.logsumexp except it accepts a ragged tensor.
+  See https://pytorch.org/docs/stable/generated/torch.logsumexp.html
```
Comment (csukuangfj) [inline k2/python/csrc/torch/v2/doc/any.h:938]:
```suggestion
  See `<https://pytorch.org/docs/stable/generated/torch.logsumexp.html>`_
```

**Review by csukuangfj — APPROVED** [review]
